# Project Title:
Income Prediction Challenge For Azubian

# Business Understanding:

The "Income Prediction Challenge for Azubian" is a data-driven initiative that seeks to address the critical issue of income inequality in developing nations. The project focuses on utilizing machine learning techniques to predict whether an individual's income falls above or below a specific income threshold. By developing a robust predictive model, we aim to contribute to more accurate and cost-effective methods of monitoring key population indicators, such as income levels, between census years. This valuable information will empower policymakers to take more informed actions to mitigate and manage income inequality on a global scale.

### 1.1. Objectives:

- **Income Prediction Model:** 
The primary goal is to create a machine learning model capable of determining whether an individual's income exceeds a specified threshold.


- **Economic Inequality Mitigation:** 
By accurately predicting income levels, the project aims to support the reduction of income inequality by providing policymakers with critical insights.


- **Cost and Accuracy Improvement:** 
This solution endeavors to improve the efficiency of income-level monitoring by offering a more cost-effective and precise method compared to traditional census methods.

### 1.2. Methodology:

To achieve the project objectives, we will follow the Cross-Industry Standard Process for Data Mining (CRISP-DM) framework, a process used to guide the machine learning lifecycle. It is a six-phase process consisting of these key phases:

**i. Business Understanding:**
- Gain a deep understanding of the problem, its significance, and the potential impact of addressing income inequality.
- Define the objectives and the F1 score evaluation metric for model performance.

**ii. Data Understanding:**
- Load the provided training, testing and variable definitions datasets.
- Explore the provided datasets, including variable descriptions.
- Analyze the Train.csv dataset, which includes target income labels, to understand the data's structure and relationships.

**iii. Data Preparation:**
- Preprocess the data by handling missing values and data anomalies.
- Perform feature engineering to prepare the data for model training.

**iv. Modeling:**
- Select and implement machine learning algorithms suitable for classification tasks.
- Train the predictive models on the Train.csv dataset using features to predict income labels.

**v. Evaluation:**
- Assess the models' performance using the F1 score, which combines precision and recall, to ensure the model's accuracy and select the best performing model.
- Perform cross validation and hyperparameter tuning.

**vi. Deployment:**
- Deploy the trained model for prediction on the Test.csv dataset, which does not include target-related columns.
- The deployment to be done through one of these frameworks; Streamlit, Gradio or FastAPI through Docker Containerization and deployment on Hugging Face. This will allow external users to interact with the application.

By following the CRISP-DM framework, we aim to create a robust income prediction model that can effectively support efforts to address income inequality, provide policymakers with valuable insights, and improve the accuracy of population monitoring between census years. This project has the potential to make a significant impact on global economic equality.

# Data Understanding

## Setup

### Installations

### Importation of Relevant Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Data Loading

### Loading the Variable Definitions, Train and Test Datasets

#### Variable Definitions Dataset

In [2]:
# Reading Variable Definitions dataset
def_df = pd.read_csv('VariableDefinitions.csv')
def_df

Column  \
0                              age   
1                           gender   
2                        education   
3                            class   
4              education_institute   
5                   marital_status   
6                             race   
7                      is_hispanic   
8            employment_commitment   
9              unemployment_reason   
10                 employment_stat   
11                   wage_per_hour   
12                  is_labor_union   
13           working_week_per_year   
14                   industry_code   
15              industry_code_main   
16                 occupation_code   
17            occupation_code_main   
18                  total_employed   
19                  household_stat   
20               household_summary   
21                 under_18_family   
22    veterans_admin_questionnaire   
23                     vet_benefit   
24                      tax_status   
25                           gains   
26                          losses   
27                   stocks_status   
28                     citizenship   
29                        mig_year   
30            country_of_birth_own   
31         country_of_birth_father   
32         country_of_birth_mother   
33    migration_code_change_in_msa   
34          migration_prev_sunbelt   
35  migration_code_move_within_reg   
36    migration_code_change_in_reg   
37            residence_1_year_ago   
38               old_residence_reg   
39             old_residence_state   
40            importance_of_record   
41              income_above_limit   

                                      Description  
0                               Age Of Individual  
1                                          Gender  
2                                       Education  
3                                 Class Of Worker  
4   Enrolled Educational Institution in last week  
5                                  Marital_Status  
6                                            Race  
7                                 Hispanic Origin  
8               Full Or Part Time Employment Stat  
9                         Reason For Unemployment  
10           Has Own Business Or Is Self Employed  
11                                  Wage Per Hour  
12                        Member Of A Labor Union  
13                         Weeks Worked In A Year  
14                                       Category  
15                            Major Industry Code  
16                                       Category  
17                          Major Occupation Code  
18                Num Persons Worked For Employer  
19             Detailed Household And Family Stat  
20        Detailed Household Summary In Household  
21                        Family Members Under 18  
22      Fill Inc Questionnaire For Veterans Admin  
23                                       Benefits  
24                               Tax Filer Status  
25                                          Gains  
26                                         Losses  
27                           Divdends From Stocks  
28                                    Citizenship  
29                                           Year  
30                     Individual's Birth Country  
31                         Father's Birth Country  
32                         Mother's Birth Country  
33                   Migration Code Change In Msa  
34                         Migration Prev Sunbelt  
35                 Migration Code Move Within Reg  
36                   Migration Code Change In Reg  
37                   Live In This House Year Ago?  
38                   Region Of Previous Residence  
39                    State Of Previous Residence  
40                             Weight Of Instance  
41                           Is income above 50k?

These variable definitions provide a clear understanding of the features and target variable used in the dataset, which is essential for data analysis and modeling.


#### Train Dataset

In [3]:
# Reading train dataset
train_df = pd.read_csv('Train.csv')
train_df.head()

ID  age   gender               education                class  \
0  ID_TZ0000   79   Female    High school graduate                  NaN   
1  ID_TZ0001   65   Female    High school graduate                  NaN   
2  ID_TZ0002   21     Male   12th grade no diploma   Federal government   
3  ID_TZ0003    2   Female                Children                  NaN   
4  ID_TZ0004   70     Male    High school graduate                  NaN   

  education_institute                    marital_status  \
0                 NaN                           Widowed   
1                 NaN                           Widowed   
2                 NaN                     Never married   
3                 NaN                     Never married   
4                 NaN   Married-civilian spouse present   

                         race is_hispanic      employment_commitment  ...  \
0                       White   All other         Not in labor force  ...   
1                       White   All other   Children or Armed Forces  ...   
2                       Black   All other   Children or Armed Forces  ...   
3   Asian or Pacific Islander   All other   Children or Armed Forces  ...   
4                       White   All other         Not in labor force  ...   

  country_of_birth_mother  migration_code_change_in_msa  \
0                      US                             ?   
1                      US                     unchanged   
2                      US                     unchanged   
3                   India                     unchanged   
4                      US                             ?   

   migration_prev_sunbelt migration_code_move_within_reg  \
0                       ?                              ?   
1                     NaN                      unchanged   
2                     NaN                      unchanged   
3                     NaN                      unchanged   
4                       ?                              ?   

   migration_code_change_in_reg  residence_1_year_ago old_residence_reg  \
0                             ?                   NaN               NaN   
1                     unchanged                  Same               NaN   
2                     unchanged                  Same               NaN   
3                     unchanged                  Same               NaN   
4                             ?                   NaN               NaN   

   old_residence_state importance_of_record  income_above_limit  
0                  NaN              1779.74         Below limit  
1                  NaN              2366.75         Below limit  
2                  NaN              1693.42         Below limit  
3                  NaN              1380.27         Below limit  
4                  NaN              1580.79         Below limit  

[5 rows x 43 columns]

#### Test Dataset

In [4]:
# Reading test dataset
test_df = pd.read_csv('Test.csv')
test_df.head()

ID  age   gender                    education     class  \
0  ID_TZ209499   54     Male         High school graduate   Private   
1  ID_TZ209500   53     Male             5th or 6th grade   Private   
2  ID_TZ209501   42     Male   Bachelors degree(BA AB BS)   Private   
3  ID_TZ209502   16   Female                    9th grade       NaN   
4  ID_TZ209503   16     Male                    9th grade       NaN   

  education_institute                    marital_status    race  \
0                 NaN   Married-civilian spouse present   White   
1                 NaN   Married-civilian spouse present   White   
2                 NaN   Married-civilian spouse present   White   
3         High school                     Never married   White   
4         High school                     Never married   White   

                  is_hispanic      employment_commitment  ...  \
0                   All other   Children or Armed Forces  ...   
1   Central or South American        Full-time schedules  ...   
2                   All other        Full-time schedules  ...   
3                   All other   Children or Armed Forces  ...   
4                   All other         Not in labor force  ...   

  country_of_birth_father  country_of_birth_mother  \
0                      US                       US   
1             El-Salvador              El-Salvador   
2                      US                       US   
3                      US                       US   
4                      US                       US   

   migration_code_change_in_msa migration_prev_sunbelt  \
0                     unchanged                    NaN   
1                             ?                      ?   
2                             ?                      ?   
3                     unchanged                    NaN   
4                             ?                      ?   

   migration_code_move_within_reg  migration_code_change_in_reg  \
0                       unchanged                     unchanged   
1                               ?                             ?   
2                               ?                             ?   
3                       unchanged                     unchanged   
4                               ?                             ?   

  residence_1_year_ago  old_residence_reg old_residence_state  \
0                 Same                NaN                 NaN   
1                  NaN                NaN                 NaN   
2                  NaN                NaN                 NaN   
3                 Same                NaN                 NaN   
4                  NaN                NaN                 NaN   

   importance_of_record  
0               3388.96  
1               1177.55  
2               4898.55  
3               1391.44  
4               1933.18  

[5 rows x 42 columns]

## Hypothesis    

**Null Hypothesis (H0):** There is no significant association between an individual's education level and the likelihood of having an income above the specified threshold.


**Alternative Hypothesis (H1):** Individuals with higher education levels are significantly more likely to have incomes above the specified threshold.


- The null hypothesis (H0) suggests that education level and income level are not related, meaning that having a higher education level does not increase the likelihood of earning an income above the threshold.

- The alternative hypothesis (H1) posits that there is a significant association between education level and income level, indicating that higher education levels are linked to a higher likelihood of having an income above the threshold.

We will conduct statistical tests to either accept or reject the null hypothesis in favor of the alternative hypothesis based on the evidence provided by the dataset.

## Key Analytical Questions:

To gain insights into the dataset and validate the hypothesis, we can formulate several key analytical questions for EDA:

i. What is the distribution of income levels in the dataset (above the threshold vs. below the threshold)?

- This question will provide an initial understanding of the balance between the two income categories.

ii. How does age relate to income levels in the dataset?

- Analyzing the age distribution among individuals with different income levels may reveal patterns related to age and income.

iii. Is there a significant gender-based income disparity?

- Exploring income levels based on gender can help us understand if gender plays a role in income categorization.

iv. What is the distribution of education levels among individuals with different income levels?

- Analyzing the educational backgrounds of individuals in both income categories can help assess the hypothesis regarding education and income.

v. Are there differences in employment status between the two income groups?

- Investigating the employment status and commitment of individuals based on income categories can provide insights into the relationship between employment and income.

vi. How do race and ethnicity correlate with income levels in the dataset?

- Understanding the distribution of income levels across different racial and ethnic groups can shed light on potential disparities.

vii. Are capital gains and losses associated with higher incomes?

- Examining the presence and amounts of capital gains and losses can help determine their impact on income levels.

viii. What is the relationship between occupation and income categories?

- Analyzing the occupation categories and their distribution among income groups can provide insights into employment roles.

ix. Do migration patterns or changes in residence relate to income levels?

- Investigating migration and residence changes may reveal how geographic mobility affects income.

x. How does tax status correspond to income levels?

Analyzing tax filing status can provide information about the impact of tax-related factors on income.

By exploring these key analytical questions, we can perform an in-depth EDA to uncover patterns, correlations, and potential factors influencing income levels. Additionally, it will help us determine whether our initial hypothesis regarding education and income holds true in this dataset.

## Exploratory Data Analysis (EDA):

### Understanding the datasets

An in-depth exploration of the datasets is presented to gain insights into the available variables,their distributions and relationships. This step will provide an initial undertanding of the datasets to identify any data quality issues that will inform the cleaning and pre-processing.

#### i. Column Information of The Datasets

In [5]:
# Column information of the train dataset
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209499 entries, 0 to 209498
Data columns (total 43 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ID                              209499 non-null  object 
 1   age                             209499 non-null  int64  
 2   gender                          209499 non-null  object 
 3   education                       209499 non-null  object 
 4   class                           104254 non-null  object 
 5   education_institute             13302 non-null   object 
 6   marital_status                  209499 non-null  object 
 7   race                            209499 non-null  object 
 8   is_hispanic                     209499 non-null  object 
 9   employment_commitment           209499 non-null  object 
 10  unemployment_reason             6520 non-null    object 
 11  employment_stat                 209499 non-null  int64  
 12  wage_per_hour   

- The **train dataset** contains **209,499 rows** and **43 columns**.

- The columns include various demographic and socioeconomic features, as well as the target variable "income_above_limit," which indicates whether the individual's income is above or below the income threshold ($50,000)


- Many columns contain missing values. For example, 'class,' 'education_institute,' 'unemployment_reason,' 'is_labor_union,' 'under_18_family,' 'veterans_admin_questionnaire,' 'old_residence_reg,' 'old_residence_state,' 'migration_code_change_in_msa,' 'migration_prev_sunbelt,' 'migration_code_move_within_reg,' and 'migration_code_change_in_reg' have significant missing values.

In [6]:
# Column information of the test dataset
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89786 entries, 0 to 89785
Data columns (total 42 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ID                              89786 non-null  object 
 1   age                             89786 non-null  int64  
 2   gender                          89786 non-null  object 
 3   education                       89786 non-null  object 
 4   class                           44707 non-null  object 
 5   education_institute             5616 non-null   object 
 6   marital_status                  89786 non-null  object 
 7   race                            89786 non-null  object 
 8   is_hispanic                     89786 non-null  object 
 9   employment_commitment           89786 non-null  object 
 10  unemployment_reason             2680 non-null   object 
 11  employment_stat                 89786 non-null  int64  
 12  wage_per_hour                   

#### ii. Shape of The Datasets

In [7]:
# The shape of the train dataset
train_df.shape

(209499, 43)

In [8]:
# The shape of the test dataset
test_df.shape

(89786, 42)

#### iii. Summary Statistics Datasets

In [9]:
# Summary Statistics of The Train Dataset
train_df.describe().round(3)

age  employment_stat  wage_per_hour  working_week_per_year  \
count  209499.000       209499.000     209499.000             209499.000   
mean       34.519            0.177         55.433                 23.159   
std        22.307            0.556        276.757                 24.398   
min         0.000            0.000          0.000                  0.000   
25%        15.000            0.000          0.000                  0.000   
50%        33.000            0.000          0.000                  8.000   
75%        50.000            0.000          0.000                 52.000   
max        90.000            2.000       9999.000                 52.000   

       industry_code  occupation_code  total_employed  vet_benefit  \
count     209499.000       209499.000      209499.000   209499.000   
mean          15.332           11.322           1.956        1.516   
std           18.050           14.461           2.365        0.851   
min            0.000            0.000           0.000        0.000   
25%            0.000            0.000           0.000        2.000   
50%            0.000            0.000           1.000        2.000   
75%           33.000           26.000           4.000        2.000   
max           51.000           46.000           6.000        2.000   

            gains      losses  stocks_status  mig_year  importance_of_record  
count  209499.000  209499.000     209499.000  209499.0            209499.000  
mean      435.927      36.882        194.533      94.5              1740.888  
std      4696.360     270.383       1956.376       0.5               995.560  
min         0.000       0.000          0.000      94.0                37.870  
25%         0.000       0.000          0.000      94.0              1061.290  
50%         0.000       0.000          0.000      94.0              1617.040  
75%         0.000       0.000          0.000      95.0              2185.480  
max     99999.000    4608.000      99999.000      95.0             18656.300

In [10]:
# Summary Statistics of The Test Dataset
test_df.describe().round(3)

age  employment_stat  wage_per_hour  working_week_per_year  \
count  89786.000        89786.000      89786.000              89786.000   
mean      34.586            0.176         54.339                 23.224   
std       22.346            0.554        265.198                 24.418   
min        0.000            0.000          0.000                  0.000   
25%       15.000            0.000          0.000                  0.000   
50%       33.000            0.000          0.000                  8.000   
75%       50.000            0.000          0.000                 52.000   
max       90.000            2.000       9400.000                 52.000   

       industry_code  occupation_code  total_employed  vet_benefit      gains  \
count      89786.000        89786.000       89786.000    89786.000  89786.000   
mean          15.377           11.298           1.956        1.518    421.978   
std           18.063           14.445           2.364        0.849   4610.516   
min            0.000            0.000           0.000        0.000      0.000   
25%            0.000            0.000           0.000        2.000      0.000   
50%            0.000            0.000           1.000        2.000      0.000   
75%           33.000           26.000           4.000        2.000      0.000   
max           51.000           46.000           6.000        2.000  99999.000   

          losses  stocks_status   mig_year  importance_of_record  
count  89786.000      89786.000  89786.000             89786.000  
mean      36.773        198.926     94.501              1738.264  
std      268.401       1893.917      0.500               990.837  
min        0.000          0.000     94.000                42.820  
25%        0.000          0.000     94.000              1059.115  
50%        0.000          0.000     95.000              1617.345  
75%        0.000          0.000     95.000              2193.735  
max     4608.000      99999.000     95.000             12960.200

#### iv. Checking for Missing Values in The Datasets

In [11]:
# Define The datasets
datasets = {'train': train_df, 'test': test_df}

# Define a function to print the missing values in the datasets
def show_missing_values(datasets):
    for name, data in datasets.items():
        title = f"Missing values in the {name.capitalize()} dataset:"
        
        # Create an underline of asterisks
        underline = '*' * len(title)  
        print(title)
        print(underline)
        
        # Calculate the sum of missing values for each column
        missing_values = data.isnull().sum()
        
        # Exclude columns with 0 as missing values
        missing_values = missing_values[missing_values > 0]
        
        # Check if there are any columns with missing values
        if not missing_values.empty:
            
            # Calculate the percentage of missing values
            missing_percentages = ((missing_values / len(data)) * 100).round(2)
            
            # Display the columns with missing values, the count of missing values, and their percentages
            missing_info = pd.DataFrame({
                'Column': missing_values.index, 
                'Missing Values': missing_values,
                'Missing Values Percentage (%)': missing_percentages
            })
            
            # Use to_string to remove index column
            print(missing_info.to_string(index=False))  
            print('===' * 26)
        else:
            # If no missing values found, indicate that
            print("No missing values found.")
        print()

# Call the function to show missing values in the datasets
show_missing_values(datasets)

Missing values in the Train dataset:
************************************
                        Column  Missing Values  Missing Values Percentage (%)
                         class          105245                          50.24
           education_institute          196197                          93.65
           unemployment_reason          202979                          96.89
                is_labor_union          189420                          90.42
          occupation_code_main          105694                          50.45
               under_18_family          151654                          72.39
  veterans_admin_questionnaire          207415                          99.01
  migration_code_change_in_msa            1588                           0.76
        migration_prev_sunbelt           88452                          42.22
migration_code_move_within_reg            1588                           0.76
  migration_code_change_in_reg            1588                      

In [12]:
# Define a function to print unique values and missing value percentages for columns with missing values
def print_unique_and_missing(dataset):
    for column in dataset.columns:
        if dataset[column].isnull().any():
            
            # Calculate the missing values percentage
            missing_percentage = (dataset[column].isnull().sum() / len(dataset)) * 100

            # Create an underline of asterisks
            underline = '*' * len(f"Column: {column}")
            print(f"Column: {column}")
            print(underline)
                        
            # Print the unique values
            unique_values = dataset[column].unique()
            print(f"Unique Values: {unique_values}")

            # Print the missing values percentage
            print(f"Missing Values Percentage: {missing_percentage:.2f}%")
            print('===' * 26)
            print()


# Print unique values and missing value percentages for columns with missing values in the 'train' dataset
print("Unique values and missing value percentages in the 'train' dataset:")
print_unique_and_missing(train_df)

Unique values and missing value percentages in the 'train' dataset:
Column: class
*************
Unique Values: [nan ' Federal government' ' Private' ' Local government'
 ' Self-employed-incorporated' ' Self-employed-not incorporated'
 ' State government' ' Without pay' ' Never worked']
Missing Values Percentage: 50.24%

Column: education_institute
***************************
Unique Values: [nan ' High school' ' College or university']
Missing Values Percentage: 93.65%

Column: unemployment_reason
***************************
Unique Values: [nan ' Other job loser' ' Re-entrant' ' Job leaver'
 ' Job loser - on layoff' ' New entrant']
Missing Values Percentage: 96.89%

Column: is_labor_union
**********************
Unique Values: [nan ' No' 'Same']
Missing Values Percentage: 90.42%

Column: occupation_code_main
****************************
Unique Values: [nan ' Adm support including clerical' ' Other service'
 ' Executive admin and managerial' ' Sales'
 ' Machine operators assmblrs & inspct

In [13]:
# Print unique values and missing value percentages for columns with missing values in the 'test' dataset
print("Unique values and missing value percentages in the 'test' dataset:")
print_unique_and_missing(test_df)

Unique values and missing value percentages in the 'test' dataset:
Column: class
*************
Unique Values: [' Private' nan ' Federal government' ' Local government'
 ' Self-employed-incorporated' ' Self-employed-not incorporated'
 ' State government' ' Never worked' ' Without pay']
Missing Values Percentage: 50.21%

Column: education_institute
***************************
Unique Values: [nan ' High school' ' College or university']
Missing Values Percentage: 93.75%

Column: unemployment_reason
***************************
Unique Values: [nan ' Re-entrant' ' Other job loser' ' New entrant'
 ' Job loser - on layoff' ' Job leaver']
Missing Values Percentage: 97.02%

Column: is_labor_union
**********************
Unique Values: [' No' nan 'Same']
Missing Values Percentage: 90.57%

Column: occupation_code_main
****************************
Unique Values: [' Transportation and material moving'
 ' Machine operators assmblrs & inspctrs' ' Professional specialty' nan
 ' Adm support including cle

##### Handling Missing Values and Leading Blank Spaces
In this section, we address missing values in the dataset. Proper handling of missing data is crucial for robust analysis and ensures data consistency.

There are several instances of '?' in different columns of the dataset. These should be replaced with NaN values since they also symbolize missing values.

We observed that some columns such as the 'class' and 'occupation_code_main' columns have leading blank spaces that may impact data analysis. To remove these spaces and replace the '?' with NaN, we use the following code:

In [14]:
def process_dataset(dataset):
    
    # Remove leading spaces from all columns in the dataset
    dataset = dataset.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

    # Replace '?' with NaN for the entire DataFrame
    dataset = dataset.replace('?', np.nan)

    return dataset

In [15]:
# Process the 'train' dataset
train_df = process_dataset(train_df)

# Process the 'test' dataset
test_df = process_dataset(test_df)

# Define a function to print unique values and missing value percentages for columns with missing values
def print_missing_values(dataset):
    for column in dataset.columns:
        if dataset[column].isnull().any():
            unique_values = dataset[column].unique()
        
            # Calculate the missing values percentage
            missing_percentage = (dataset[column].isnull().sum() / len(dataset)) * 100
        
            # Create an underline of asterisks
            underline = '*' * len(f"Column: {column}")
            print(f"Column: {column}")
            print(underline)
        
            # Print the unique values
            print(f"Unique Values: {unique_values}")
        
            # Print the missing values percentage  
            print(f"Missing Values Percentage: {missing_percentage:.2f}%")
            print('===' * 29)
            print()

In [16]:
# Print unique values and missing value percentages for columns with missing values in the 'train' dataset
print_missing_values(train_df)

Column: class
*************
Unique Values: [nan 'Federal government' 'Private' 'Local government'
 'Self-employed-incorporated' 'Self-employed-not incorporated'
 'State government' 'Without pay' 'Never worked']
Missing Values Percentage: 50.24%

Column: education_institute
***************************
Unique Values: [nan 'High school' 'College or university']
Missing Values Percentage: 93.65%

Column: unemployment_reason
***************************
Unique Values: [nan 'Other job loser' 'Re-entrant' 'Job leaver' 'Job loser - on layoff'
 'New entrant']
Missing Values Percentage: 96.89%

Column: is_labor_union
**********************
Unique Values: [nan 'No' 'Same']
Missing Values Percentage: 90.42%

Column: occupation_code_main
****************************
Unique Values: [nan 'Adm support including clerical' 'Other service'
 'Executive admin and managerial' 'Sales'
 'Machine operators assmblrs & inspctrs'
 'Precision production craft & repair' 'Professional specialty'
 'Handlers equip clea

In [17]:
# Print unique values and missing value percentages for columns with missing values in the 'test' dataset
print_missing_values(test_df)

Column: class
*************
Unique Values: ['Private' nan 'Federal government' 'Local government'
 'Self-employed-incorporated' 'Self-employed-not incorporated'
 'State government' 'Never worked' 'Without pay']
Missing Values Percentage: 50.21%

Column: education_institute
***************************
Unique Values: [nan 'High school' 'College or university']
Missing Values Percentage: 93.75%

Column: unemployment_reason
***************************
Unique Values: [nan 'Re-entrant' 'Other job loser' 'New entrant' 'Job loser - on layoff'
 'Job leaver']
Missing Values Percentage: 97.02%

Column: is_labor_union
**********************
Unique Values: ['No' nan 'Same']
Missing Values Percentage: 90.57%

Column: occupation_code_main
****************************
Unique Values: ['Transportation and material moving'
 'Machine operators assmblrs & inspctrs' 'Professional specialty' nan
 'Adm support including clerical' 'Precision production craft & repair'
 'Technicians and related support' 'Execut

In [18]:
# Define a threshold percentage for missing values
threshold = 0.70

# Create an empty list to store columns that exceed the threshold
train_columns_to_drop = []
test_columns_to_drop = []
         
            
# Iterate through each column in the train dataset           
for col in train_df.columns:

    # Calculate the percentage of missing values for the current column
    missing_percentage = train_df[col].isnull().sum() / len(train_df)
    
    # Check if the missing percentage is greater than the defined threshold
    if missing_percentage > threshold:
        
        # If it exceeds the threshold, add the column name to the list      
        train_columns_to_drop.append(col)
        

# Iterate through each column in the test dataset
for col in test_df.columns:
    
    # Calculate the percentage of missing values for the current column
    missing_percentage = test_df[col].isnull().sum() / len(test_df)
    
    # Check if the missing percentage is greater than the defined threshold
    if missing_percentage > threshold:
        
        # If it exceeds the threshold, add the column name to the list            
        test_columns_to_drop.append(col)

# Drop the identified columns with high missing values from both the train and test datasets
train_df.drop(columns=train_columns_to_drop, inplace=True)
test_df.drop(columns=test_columns_to_drop, inplace=True)

print(f"Dropped columns in train: {train_columns_to_drop}")
print()
print(f"Dropped columns in test: {test_columns_to_drop}")

Dropped columns in train: ['education_institute', 'unemployment_reason', 'is_labor_union', 'under_18_family', 'veterans_admin_questionnaire', 'migration_prev_sunbelt', 'old_residence_reg', 'old_residence_state']

Dropped columns in test: ['education_institute', 'unemployment_reason', 'is_labor_union', 'under_18_family', 'veterans_admin_questionnaire', 'migration_prev_sunbelt', 'old_residence_reg', 'old_residence_state']


In [19]:
# Define The datasets
datasets = {'train': train_df, 'test': test_df}

# Define a function to print the missing values in the datasets
def show_missing_values(datasets):
    for name, data in datasets.items():
        title = f"Missing values in the {name.capitalize()} dataset:"
        
        # Create an underline of asterisks
        underline = '*' * len(title)  
        print(title)
        print(underline)
        
        # Calculate the sum of missing values for each column
        missing_values = data.isnull().sum()
        
        # Exclude columns with 0 as missing values
        missing_values = missing_values[missing_values > 0]
        
        # Check if there are any columns with missing values
        if not missing_values.empty:
            
            # Calculate the percentage of missing values
            missing_percentages = ((missing_values / len(data)) * 100).round(2)
            
            # Display the columns with missing values, the count of missing values, and their percentages
            missing_info = pd.DataFrame({
                'Column': missing_values.index, 
                'Missing Values': missing_values,
                'Missing Values Percentage (%)': missing_percentages
            })
            
            # Use to_string to remove index column
            print(missing_info.to_string(index=False))  
            print('===' * 26)
        else:
            # If no missing values found, indicate that
            print("No missing values found.")
        print()

# Call the function to show missing values in the datasets
show_missing_values(datasets)

Missing values in the Train dataset:
************************************
                        Column  Missing Values  Missing Values Percentage (%)
                         class          105245                          50.24
          occupation_code_main          105694                          50.45
          country_of_birth_own            3667                           1.75
       country_of_birth_father            7181                           3.43
       country_of_birth_mother            6516                           3.11
  migration_code_change_in_msa          106284                          50.73
migration_code_move_within_reg          106284                          50.73
  migration_code_change_in_reg          106284                          50.73
          residence_1_year_ago          106284                          50.73

Missing values in the Test dataset:
***********************************
                        Column  Missing Values  Missing Values Percentage

##### `class` and `occupation_code_main`

- **Class:** This column represents an individual's working class and has a missing value percentage of approximately 50.24%. The class of worker indicates whether a person is employed in various sectors, including the federal government, private sector, local government, state government, self-employed, or if they have never been employed or are working without pay. This column is crucial for analysis as an individual's worker class may significantly impact their income. Therefore, it is essential to retain and further analyze this column to develop an appropriate strategy for handling the missing values.


- **Occupation Code Main:** This column denotes the major occupation code of an individual and also has a missing values percentage of around 50.45%.


- **Handling Strategy:** Given the substantial percentage of missing values, we can consider imputing these gaps. One approach is to replace missing values with the most common class or occupation in each column (mode imputation).

In [20]:
# Handling the missing values in the 'Class' Column by observing neighbouring columns
train_df["class"].unique()

array([nan, 'Federal government', 'Private', 'Local government',
       'Self-employed-incorporated', 'Self-employed-not incorporated',
       'State government', 'Without pay', 'Never worked'], dtype=object)

In [21]:
# Display the rows with null values in the 'class' column
null_class = train_df[train_df["class"].isnull()]
null_class

ID  age  gender                   education class  \
0        ID_TZ0000   79  Female        High school graduate   NaN   
1        ID_TZ0001   65  Female        High school graduate   NaN   
3        ID_TZ0003    2  Female                    Children   NaN   
4        ID_TZ0004   70    Male        High school graduate   NaN   
8        ID_TZ0008   73  Female           7th and 8th grade   NaN   
...            ...  ...     ...                         ...   ...   
209488   ID_TZ9999   76  Female  Some college but no degree   NaN   
209492  ID_TZ99993   44  Female    1st 2nd 3rd or 4th grade   NaN   
209494  ID_TZ99995   16  Female           7th and 8th grade   NaN   
209495  ID_TZ99996   78    Male  Bachelors degree(BA AB BS)   NaN   
209496  ID_TZ99997    8  Female                    Children   NaN   

                         marital_status                       race  \
0                               Widowed                      White   
1                               Widowed                      White   
3                         Never married  Asian or Pacific Islander   
4       Married-civilian spouse present                      White   
8       Married-civilian spouse present                      Black   
...                                 ...                        ...   
209488                    Never married                      White   
209492  Married-civilian spouse present                      White   
209494                    Never married                      White   
209495  Married-civilian spouse present                      White   
209496                    Never married                      White   

               is_hispanic     employment_commitment  employment_stat  ...  \
0                All other        Not in labor force                0  ...   
1                All other  Children or Armed Forces                0  ...   
3                All other  Children or Armed Forces                0  ...   
4                All other        Not in labor force                0  ...   
8                All other        Not in labor force                0  ...   
...                    ...                       ...              ...  ...   
209488           All other        Not in labor force                0  ...   
209492  Mexican (Mexicano)        Not in labor force                0  ...   
209494           All other        Not in labor force                0  ...   
209495           All other        Not in labor force                0  ...   
209496           All other  Children or Armed Forces                0  ...   

        mig_year  country_of_birth_own  country_of_birth_father  \
0             95                    US                       US   
1             94                    US                       US   
3             94                    US                    India   
4             95                    US                       US   
8             95                    US                       US   
...          ...                   ...                      ...   
209488        95                    US                      NaN   
209492        95                Mexico                   Mexico   
209494        95                    US                       US   
209495        95                    US                       US   
209496        95                    US                       US   

       country_of_birth_mother  migration_code_change_in_msa  \
0                           US                           NaN   
1                           US                     unchanged   
3                        India                     unchanged   
4                           US                           NaN   
8                           US                           NaN   
...                        ...                           ...   
209488                     NaN                           NaN   
209492                  Mexico                           NaN   
209494                      

In [22]:
# Display the rows with participants under the age of 18
children_df = train_df[train_df["age"]<=18]
children_df

ID  age  gender          education class marital_status  \
3        ID_TZ0003    2  Female           Children   NaN  Never married   
11       ID_TZ0011    4    Male           Children   NaN  Never married   
13       ID_TZ0013   16  Female          9th grade   NaN  Never married   
16       ID_TZ0016    5  Female           Children   NaN  Never married   
29       ID_TZ0029   14    Male           Children   NaN  Never married   
...            ...  ...     ...                ...   ...            ...   
209477   ID_TZ9998    8    Male           Children   NaN  Never married   
209481  ID_TZ99983   11    Male           Children   NaN  Never married   
209483  ID_TZ99985   11    Male           Children   NaN  Never married   
209494  ID_TZ99995   16  Female  7th and 8th grade   NaN  Never married   
209496  ID_TZ99997    8  Female           Children   NaN  Never married   

                             race       is_hispanic     employment_commitment  \
3       Asian or Pacific Islander         All other  Children or Armed Forces   
11                          Black         All other  Children or Armed Forces   
13                          White         All other  Children or Armed Forces   
16                          White         All other  Children or Armed Forces   
29                          White         All other  Children or Armed Forces   
...                           ...               ...                       ...   
209477                      Black         All other  Children or Armed Forces   
209481                      Black         All other  Children or Armed Forces   
209483                      White  Mexican-American  Children or Armed Forces   
209494                      White         All other        Not in labor force   
209496                      White         All other  Children or Armed Forces   

        employment_stat  ...  mig_year  country_of_birth_own  \
3                     0  ...        94                    US   
11                    0  ...        94                    US   
13                    0  ...        94                    US   
16                    0  ...        95                    US   
29                    0  ...        95                    US   
...                 ...  ...       ...                   ...   
209477                0  ...        95                    US   
209481                0  ...        94                    US   
209483                0  ...        95                    US   
209494                0  ...        95                    US   
209496                0  ...        95                    US   

        country_of_birth_father country_of_birth_mother  \
3                         India                   India   
11                           US                      US   
13                           US                      US   
16                           US                      US   
29                           US                      US   
...                         ...                     ...   
209477                       US                      US   
209481                       US                      US   
209483                Nicaragua                  Mexico   
209494                       US                      US   
209496                       US                      US   

        migration_code_change_in_msa migration_code_move_within_reg  \
3                          unchanged                      unchanged   
11                         unchanged                      unchanged   
13                         unchanged                      unchanged   
16                               NaN                            NaN   
29                               NaN                            NaN   
...                              ...                            ...   
209477                           NaN                            NaN   
209481                    MSA to MSA       Different state in South   
209483                   

In [23]:
# Display the rows for participants with 'Never worked' as their class in the 'class' column
never_worked = train_df[train_df["class"] == "Never worked"]
never_worked

ID  age  gender                               education  \
2281    ID_TZ101153   38    Male  Masters degree(MA MS MEng MEd MSW MBA)   
2375    ID_TZ101238   28    Male              Some college but no degree   
2394    ID_TZ101255   16  Female                               9th grade   
3702    ID_TZ102433   16  Female                              11th grade   
3853     ID_TZ10257   16  Female                               9th grade   
...             ...  ...     ...                                     ...   
202738   ID_TZ93853   23  Female                       7th and 8th grade   
202869   ID_TZ93972   16    Male                               9th grade   
203421   ID_TZ94474   24    Male                              11th grade   
208257   ID_TZ98870   17  Female                              11th grade   
208596   ID_TZ99179   21  Female              Bachelors degree(BA AB BS)   

               class marital_status   race                is_hispanic  \
2281    Never worked  Never married  White                  All other   
2375    Never worked  Never married  White                  All other   
2394    Never worked  Never married  White                  All other   
3702    Never worked  Never married  White                  All other   
3853    Never worked  Never married  White           Mexican-American   
...              ...            ...    ...                        ...   
202738  Never worked  Never married  White           Mexican-American   
202869  Never worked  Never married  White                  All other   
203421  Never worked  Never married  White                  All other   
208257  Never worked  Never married  Black                  All other   
208596  Never worked  Never married  White  Central or South American   

           employment_commitment  employment_stat  ...  mig_year  \
2281    Children or Armed Forces                0  ...        94   
2375    Children or Armed Forces                0  ...        94   
2394    Children or Armed Forces                0  ...        94   
3702    Children or Armed Forces                0  ...        94   
3853       Unemployed part- time                0  ...        95   
...                          ...              ...  ...       ...   
202738      Unemployed full-time                0  ...        95   
202869  Children or Armed Forces                0  ...        94   
203421      Unemployed full-time                0  ...        95   
208257  Children or Armed Forces                0  ...        94   
208596  Children or Armed Forces                0  ...        94   

        country_of_birth_own  country_of_birth_father country_of_birth_mother  \
2281                      US                       US                      US   
2375                      US                       US                      US   
2394                      US                       US                      US   
3702                      US                       US                      US   
3853                      US                   Mexico                      US   
...                      ...                      ...                     ...   
202738                    US                       US                      US   
202869                    US                       US                      US   
203421                    US                       US                      US   
208257                    US                       US                      US   
208596                   NaN                      NaN                     NaN   

        migration_code_change_in_msa migration_code_move_within_reg  \
2281                       unchanged                      unchanged   
2375                       unchanged                      unchanged   
2394                       unchanged                      unchanged   
3702                       unchanged                      unchanged   
3853                             NaN                            NaN   
...        

When examining the rows with missing values in the 'Class' column, it appears that these individuals are primarily young (below 18), with 'Children' listed as their education status. This suggests that most of the missing values in the 'Class' column are due to individuals who are children or not yet part of the labor force due to their age.

It's possible that the missing values in the 'Class' column are related to children or individuals not currently employed. These cases are labeled as 'Children' in the 'Education' column and 'Not in labor force' or 'Children or Armed Forces' in the 'Employment Commitment' column.

Based on this observation, the imputation strategy for the missing values in the 'Class' column where the participants age is below 18 is assigning the value 'Never worked' which is part of the unique values in the class column.

In [24]:
# Impute missing 'Class' values for individuals below 18 with 'Never worked'
train_df.loc[train_df['age'] < 18, 'class'] = 'Never worked'
test_df.loc[test_df['age'] < 18, 'class'] = 'Never worked'

We have have handled the missing values in the 'class' colum of both datasets where the partcipants are aged below 18.

Next we look at the remaining missing values in the 'class' columns and impute them with the most frequent class.

In [25]:
train_df["class"].value_counts()

Private                           73388
Never worked                      58780
Self-employed-not incorporated     8827
Local government                   8001
State government                   4459
Self-employed-incorporated         3404
Federal government                 3035
Without pay                         146
Name: class, dtype: int64

The most frequent class is **Private**

In [26]:
# Define the imputation value
imputation_value = train_df['class'].mode()[0]

# Impute missing 'class' values with the most frequent class in the 'train' dataset
train_df['class'].fillna(imputation_value, inplace=True)

# Impute missing 'class' values with the most frequent class in the 'test' dataset
test_df['class'].fillna(imputation_value, inplace=True)

In [27]:
# Verify the imputation in both DataFrames
for name, df in datasets.items():
    missing_values_after_imputation = df['class'].isnull().sum()
    print(f"Missing values in 'class' column for {name.capitalize()} after imputation: {missing_values_after_imputation}")

Missing values in 'class' column for Train after imputation: 0
Missing values in 'class' column for Test after imputation: 0


In [28]:
# Handling Missing Values for 'occupation_code_main' column by observing neighbouring columns
train_df["occupation_code_main"].unique()

array([nan, 'Adm support including clerical', 'Other service',
       'Executive admin and managerial', 'Sales',
       'Machine operators assmblrs & inspctrs',
       'Precision production craft & repair', 'Professional specialty',
       'Handlers equip cleaners etc',
       'Transportation and material moving',
       'Farming forestry and fishing', 'Private household services',
       'Technicians and related support', 'Protective services',
       'Armed Forces'], dtype=object)

- The column "occupation_code_main" represents the major occupation code of individuals, and it has various unique values that describe different occupation categories. 


- These values include 'Adm support including clerical,' 'Sales,' 'Machine operators assmblrs & inspctrs,' 'Professional specialty,' 'Transportation and material moving,' and more. 


- This column is essential for understanding the occupation of individuals in the dataset because it indicates the occupation of an individual which might have a relationship with their income since some occupations tend to pay more than others.

In [29]:
train_df["occupation_code"].unique()

array([ 0, 26, 31,  2, 17, 36, 33,  8, 10, 40, 29, 39, 23, 34, 43, 38,  3,
       35, 19, 25, 27, 24, 30,  1, 32, 15,  5, 28, 37, 12,  4, 16,  7, 14,
       44, 18, 13,  9, 42, 21, 41, 22, 11, 45,  6, 20, 46], dtype=int64)

In [30]:
# Check missing values in 'occupation_code_main' column before imputation
missing_values_before = train_df['occupation_code_main'].isnull().sum()
print(f"Missing values in 'occupation_code_main' before imputation: {missing_values_before}")

Missing values in 'occupation_code_main' before imputation: 105694


In [37]:
# Display the rows with null values in the 'class' column
null_occupation = train_df[train_df["occupation_code_main"].isnull()]
null_occupation

ID  age  gender                   education         class  \
0        ID_TZ0000   79  Female        High school graduate       Private   
1        ID_TZ0001   65  Female        High school graduate       Private   
3        ID_TZ0003    2  Female                    Children  Never worked   
4        ID_TZ0004   70    Male        High school graduate       Private   
8        ID_TZ0008   73  Female           7th and 8th grade       Private   
...            ...  ...     ...                         ...           ...   
209488   ID_TZ9999   76  Female  Some college but no degree       Private   
209492  ID_TZ99993   44  Female    1st 2nd 3rd or 4th grade       Private   
209494  ID_TZ99995   16  Female           7th and 8th grade  Never worked   
209495  ID_TZ99996   78    Male  Bachelors degree(BA AB BS)       Private   
209496  ID_TZ99997    8  Female                    Children  Never worked   

                         marital_status                       race  \
0                               Widowed                      White   
1                               Widowed                      White   
3                         Never married  Asian or Pacific Islander   
4       Married-civilian spouse present                      White   
8       Married-civilian spouse present                      Black   
...                                 ...                        ...   
209488                    Never married                      White   
209492  Married-civilian spouse present                      White   
209494                    Never married                      White   
209495  Married-civilian spouse present                      White   
209496                    Never married                      White   

               is_hispanic     employment_commitment  employment_stat  ...  \
0                All other        Not in labor force                0  ...   
1                All other  Children or Armed Forces                0  ...   
3                All other  Children or Armed Forces                0  ...   
4                All other        Not in labor force                0  ...   
8                All other        Not in labor force                0  ...   
...                    ...                       ...              ...  ...   
209488           All other        Not in labor force                0  ...   
209492  Mexican (Mexicano)        Not in labor force                0  ...   
209494           All other        Not in labor force                0  ...   
209495           All other        Not in labor force                0  ...   
209496           All other  Children or Armed Forces                0  ...   

        gains  losses  stocks_status                         citizenship  \
0           0       0            292                              Native   
1           0       0              0                              Native   
3           0       0              0                              Native   
4           0       0              0                              Native   
8           0       0              0                              Native   
...       ...     ...            ...                                 ...   
209488      0       0           2744                              Native   
209492      0       0              0  Foreign born- Not a citizen of U S   
209494      0       0              0                              Native   
209495      0       0             71                              Native   
209496      0       0              0                              Native   

        mig_year country_of_birth_own  country_of_birth_father  \
0             95                   US                       US   
1             94                   US                       US   
3             94                   US                    India   
4             95                   US                       US   
8             95                   US                       US   
... 

In [31]:
# Group by unique combinations of 'education' and 'occupation_code_main'
occupation_education_groups = train_df.dropna(subset=['occupation_code_main', 'education']).groupby(['education', 'occupation_code_main'])

# Calculate the mode 'occupation_code_main' for each group
mode_occupation_by_education = occupation_education_groups['occupation_code_main'].apply(lambda x: x.mode().iloc[0])

# Use the calculated modes to fill missing 'occupation_code_main' values
train_df['occupation_code_main'].fillna(train_df['education'].map(mode_occupation_by_education), inplace=True)

In [32]:
# Check missing values in 'occupation_code_main' column after imputation
missing_values_after = train_df['occupation_code_main'].isnull().sum()
print(f"Missing values in 'occupation_code_main' after imputation: {missing_values_after}")

Missing values in 'occupation_code_main' after imputation: 105694


##### `country_of_birth_own`, `country_of_birth_father`, and `country_of_birth_mother`

- **Handling Strategy:** Given that these columns have a relatively low percentage of missing values (approximately 1.66%, 3.30%, and 2.98% respectively), we can consider imputing the missing values using mode imputation. This means replacing the missing values in each column with the most common value observed in that specific column. Mode imputation is suitable for these columns as it retains the existing distribution of values while filling in the gaps.

In [33]:
# Columns to impute using mode imputation
columns_to_impute = ['country_of_birth_own', 'country_of_birth_father', 'country_of_birth_mother']

# Loop through the specified columns and perform mode imputation for the train dataset
for column in columns_to_impute:
    mode_value_train = train_df[column].mode()[0]
    train_df[column].fillna(mode_value_train, inplace=True)

# Perform mode imputation for the test dataset
for column in columns_to_impute:
    mode_value_test = test_df[column].mode()[0]
    test_df[column].fillna(mode_value_test, inplace=True)    

##### `migration_code_change_in_msa`, `migration_code_move_within_reg`, `migration_code_change_in_reg`, and `residence_1_year_ago`

- These columns have a relatively high percentage of missing values (around 50.73 for the train_df and 50.78% for the test_df).
- **Handling Strategy:** We need to check whether the columns are relevant for our analysis before going ahead with an appropriate imputation strategy.

In [34]:
# Assess the relevance of the columns
relevant_columns = ['migration_code_change_in_msa', 'migration_code_move_within_reg', 'migration_code_change_in_reg', 'residence_1_year_ago']

for column in relevant_columns:
    # Check unique values to assess relevance
    unique_values = train_df[column].unique()
    print(f"Unique values for '{column}': {unique_values}")
    print('===' * 26)
    print()

Unique values for 'migration_code_change_in_msa': [nan 'unchanged' 'MSA to MSA' 'NonMSA to nonMSA' 'MSA to nonMSA'
 'Not identifiable' 'NonMSA to MSA' 'Abroad to MSA' 'Abroad to nonMSA']

Unique values for 'migration_code_move_within_reg': [nan 'unchanged' 'Same county' 'Different state in South'
 'Different county same state' 'Different state in West'
 'Different state in Northeast' 'Abroad' 'Different state in Midwest']

Unique values for 'migration_code_change_in_reg': [nan 'unchanged' 'Same county' 'Different state same division'
 'Different county same state' 'Different region' 'Abroad'
 'Different division same region']

Unique values for 'residence_1_year_ago': [nan 'Same' 'No']



In [35]:
# List of columns to drop
columns_to_drop = ['migration_code_change_in_msa', 'migration_code_move_within_reg', 'migration_code_change_in_reg', 'residence_1_year_ago']

# Drop the specified columns from the train DataFrame
train_df.drop(columns=columns_to_drop, inplace=True)

# Drop the specified columns from the test DataFrame
test_df.drop(columns=columns_to_drop, inplace=True)

- The decision to drop the columns 'migration_code_change_in_msa,' 'migration_code_move_within_reg,' 'migration_code_change_in_reg,' and 'residence_1_year_ago' is supported by the unique values observed in these columns.

- Given the high percentage of missing values and the nature of the unique values, it is unlikely that these columns will significantly impact the prediction of an individual's income.

- The unique values are either missing, indicating an unchanged status, or showing various types of transitions in residence, which may not provide meaningful information for predicting income.

- Consequently, removing these columns is a logical step to streamline the dataset and enhance the accuracy and interpretability of the analysis or predictive model.

In [36]:
train_df

ID  age  gender                   education  \
0        ID_TZ0000   79  Female        High school graduate   
1        ID_TZ0001   65  Female        High school graduate   
2        ID_TZ0002   21    Male       12th grade no diploma   
3        ID_TZ0003    2  Female                    Children   
4        ID_TZ0004   70    Male        High school graduate   
...            ...  ...     ...                         ...   
209494  ID_TZ99995   16  Female           7th and 8th grade   
209495  ID_TZ99996   78    Male  Bachelors degree(BA AB BS)   
209496  ID_TZ99997    8  Female                    Children   
209497  ID_TZ99998   26  Female  Some college but no degree   
209498  ID_TZ99999   19    Male                  11th grade   

                     class                   marital_status  \
0                  Private                          Widowed   
1                  Private                          Widowed   
2       Federal government                    Never married   
3             Never worked                    Never married   
4                  Private  Married-civilian spouse present   
...                    ...                              ...   
209494        Never worked                    Never married   
209495             Private  Married-civilian spouse present   
209496        Never worked                    Never married   
209497             Private  Married-civilian spouse present   
209498    Local government                    Never married   

                             race is_hispanic     employment_commitment  \
0                           White   All other        Not in labor force   
1                           White   All other  Children or Armed Forces   
2                           Black   All other  Children or Armed Forces   
3       Asian or Pacific Islander   All other  Children or Armed Forces   
4                           White   All other        Not in labor force   
...                           ...         ...                       ...   
209494                      White   All other        Not in labor force   
209495                      White   All other        Not in labor force   
209496                      White   All other  Children or Armed Forces   
209497                      White   All other  Children or Armed Forces   
209498                      Black   All other  Children or Armed Forces   

        employment_stat  ...  gains  losses  stocks_status citizenship  \
0                     0  ...      0       0            292      Native   
1                     0  ...      0       0              0      Native   
2                     0  ...      0       0              0      Native   
3                     0  ...      0       0              0      Native   
4                     0  ...      0       0              0      Native   
...                 ...  ...    ...     ...            ...         ...   
209494                0  ...      0       0              0      Native   
209495                0  ...      0       0             71      Native   
209496                0  ...      0       0              0      Native   
209497                0  ...      0       0              0      Native   
209498                0  ...      0       0              0      Native   

        mig_year country_of_birth_own  country_of_birth_father  \
0             95                   US                       US   
1             94                   US                       US   
2             94                   US                       US   
3             94                   US                    India   
4             95                   US                       US   
...          ...                  ...                      ...   
209494        95                   US                       US   
209495        95                   US                       US   
209496        95                   US                       US   
209497        94                   US               